internal/gitea/repo.go

```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// parseReadme is a helper function to fetch the README from a Gitea repository and return
+// its contents as a string.
+func (r *Repository) parseReadme() error {
+	var readmeNames = []string {"README.md", "README.rst"}
+	var giteaReadme []byte
+	for _, rName := range readmeNames { 
+		var err error
+		var fileName string
+		if r.folder == "" {
+			fileName = rName
+		} else {
+			fileName = fmt.Sprintf("%s/%s", r.folder, rName)
+		}
+		giteaReadme, _, err = r.client.GetFile(r.org, r.Details.Name, r.defaultBranch, fileName, false)
+		if err == nil {
+			break
+		}
+	}
+	if len(giteaReadme) == 0 {
+		return errors.New("error getting README for repo")
+	}
+
+	content := string(giteaReadme)
+
+	// Parse contents of README to identify associated snaps, charms.
+	readme := &repos.Readme{Body: content}
+	ctx := context.Background()
+	r.Details.Snap = readme.LinkedSnap(ctx)
+	r.Details.Charm = readme.LinkedCharm(ctx)
+
+	return nil
+}
+
+// processReleases fetches a repository's releases from Gitea, then populates r.Details.Releases
+// with the information in the relevant format for releasegen.
+func (r *Repository) processReleases() error {
+	// This currently gets all releases across the entire repository, even in a
+	// monorepo. It's not clear what happens with releases in a Gitea monorepo -
+	// for example, are they not used at all? Do they all start with the name of
+	// the 'subrepo'?
+
+	isDraft := false
+	isPreRelease := false
+	opts := gitea.ListReleasesOptions{ListOptions: gitea.ListOptions{PageSize: giteaReleasesPerRepo}, IsDraft: &isDraft, IsPreRelease: &isPreRelease}
+
+	releases, _, err := r.client.ListReleases(r.org, r.Details.Name, opts)
+	if err != nil {
+		return errors.New("error listing releases for repo")
+	}
+
+	for _, rel := range releases {
+		r.Details.Releases = append(r.Details.Releases, &repos.Release{
+			ID:         rel.ID,
+			Version:    rel.TagName,
+			Timestamp:  rel.PublishedAt.Unix(),
+			Title:      rel.Title,
+			Body:       renderReleaseBody(rel.Note, r),
+			URL:        rel.URL,
+			CompareURL: fmt.Sprintf("%s/compare/%s...%s", r.Details.URL, rel.TagName, r.defaultBranch),
+		})
+	}
+
+	return nil
+}
+
+// processCommitsSinceRelease calculates the number of commits that have occurred on the default
+// branch of the repository since the last release, and populates the information in r.Details.
+func (r *Repository) processCommitsSinceRelease() error {
+	// This does not currently handle monorepos - see note about releases in
+	// `processReleases`. If there are releases in Gitea with a monorepo, then
+	// this function maybe needs to restrict the commits to ones where the tree
+	// overlaps - but maybe we actually want to know how many commits overall,
+	// because we want to count things like common code being adjusted.
+
+	// Add the commit delta between last release and default branch.
+	latestRelease := r.Details.Releases[len(r.Details.Releases)-1]
```

Comment:
Can `.Releases` ever be zero-length? In which case this lookup would panic, so if so we would want to check `len(r.Details.Releases) > 0`.